src/main/java/be/appwise/core/networking/proxyman/ProxyManNetworkDiscoveryManager.kt

```diff
@@ -0,0 +1,273 @@
+package be.appwise.core.networking.proxyman
+
+import android.content.Context
+import android.net.nsd.NsdManager
+import android.net.nsd.NsdServiceInfo
+import com.google.gson.GsonBuilder
+import com.orhanobut.logger.Logger
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import java.io.ByteArrayOutputStream
+import java.io.IOException
+import java.net.InetAddress
+import java.net.Socket
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+import java.util.zip.GZIPOutputStream
+
+
+object ProxyManNetworkDiscoveryManager {
+
+    const val MaximumSizePackage = 52428800
+
+    val proxyManGson = GsonBuilder()
+        .registerTypeAdapter(ByteArray::class.java, Base64ArrayTypeAdapter())
+        .registerTypeAdapter(Data::class.java, Base64TypeAdapter())
+        .create()
+
+    private lateinit var nsdManager: NsdManager
+
+    //port and tag that bagle uses
+    private const val SERVICE_TYPE: String = "_Proxyman._tcp."
+    private const val PORT = 43434
+    private const val TAG = "ProxyManNetworkDiscoveryManager"
+
+    private var mServiceName: String? = null
+
+    private val registrationListener = object : NsdManager.RegistrationListener {
+
+        override fun onServiceRegistered(NsdServiceInfo: NsdServiceInfo) {
+            // Save the service name. Android may have changed it in order to
+            // resolve a conflict, so update the name you initially requested
+            // with the name Android actually used.
+            mServiceName = NsdServiceInfo.serviceName
+        }
+
+        override fun onRegistrationFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
+            // Registration failed! Put debugging code here to determine why.
+            Logger.d("registrationfailed errorcode $errorCode")
+        }
+
+        override fun onServiceUnregistered(arg0: NsdServiceInfo) {
+            // Service has been unregistered. This only happens when you call
+            // NsdManager.unregisterService() and pass in this listener.
+            nsdManager.unregisterService(this)
+        }
+
+        override fun onUnregistrationFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
+            // Unregistration failed. Put debugging code here to determine why.
+            Logger.d("unregistrationfailed errorcode $errorCode")
+        }
+    }
+
+    fun teardown() {
```

Comment:
`teardown` is never used. You could use it in `registerService` before you do anything else, just in case something is already instantiated before. (Take a look at `BagelNetworkDiscoveryManager` as an example)